src/core/channel/client_uchannel.c

```diff
@@ -0,0 +1,573 @@
+/*
+ *
+ * Copyright 2015, Google Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "src/core/channel/client_uchannel.h"
+
+#include <string.h>
+
+#include "src/core/census/grpc_filter.h"
+#include "src/core/channel/channel_args.h"
+#include "src/core/channel/client_channel.h"
+#include "src/core/channel/compress_filter.h"
+#include "src/core/iomgr/iomgr.h"
+#include "src/core/support/string.h"
+#include "src/core/surface/channel.h"
+#include "src/core/transport/connectivity_state.h"
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+#include <grpc/support/useful.h>
+
+/** Microchannel (uchannel) implementation: a lightweight channel without any
```

Comment:
So... I think I want to halve the size of this file before submission to give that lightweight claim some teeth. This will likely involve some refactoring :)

Reply:
Hi, I am analyzing how channel  works in grpc c core. 

we have polling island to manage epoll event set. epoll is one of multiplexing techniques for concurrency used in Linux. In other platform, like windows, other techniques like completion port/queue is used. instead. I am familiar with epoll, kqueue but not I/O completion port/queue.

> What is purpose of completion queue and how it is implemented?

In gRPC, I noticed that, both technologies are used. we already have Endpoint to wrap fd (event). Why should we have concepts of "pollset_set" and "channel". 

A typical epoll based concurrent server looks like:

~~~ c++
...
#include <epoll.h>
#include <pthread.h>

// L176 ~ 205
struct polling_island; // wraps an "epoll" instance fd created by "epoll_create1"; the fd is exactly target what "epoll_wait" inspect with.

// epoll main loop operation, ev_epollsig_linux.c L1273 ~ 1292
static void pollset_work_and_unlock(Context *ctx, pollset *pollset, Worker *workers, ...)
{
...
ep_rv =
      epoll_pwait(epoll_fd, ep_ev, GRPC_EPOLL_MAX_EVENTS, timeout_ms, sig_mask);
if (ep_rv < 0) { ... }
...
  for (int i = 0; i < ep_rv; ++i) {
...
	// calling grpc_lfev_set_ready(exec_ctx, &fd->closure, "eventName") inside, which defined in "lockfree_event.h"
  }
}

~~~

Each operation will be executed in an independed thread. It is strange for me to have "pollset_set" and "channel".

I am familiar with golang like implementation of "channel". For now, I don't deem them as the same thing. 

Could you give more hints about grpc core logics.

I am also writing a report about how grpc is implemented and try to reproduce a simple concurrent RPC server WHICH inherits logics and spirits of gRPC.

I wish you can clearify my doubts to understand gRPC C CORE deep.